### PR TITLE
[AERIE-1605] Add plan_dataset trigger tests

### PR DESCRIPTION
**Pull request by dyst5422**
Add a test for each of the plan_dataset triggers in our SQL files

* **Tickets addressed:** AERIE-1605
* **Review:** By commit
* **Merge strategy:** Merge (no squash)

## Description
Simple SQL manipulation based tests.

## Verification
N/A

## Documentation
N/A - minor code comments added to SQL files for clarity

## Future work
None (hopefully)

### CODE REVIEW

**Review by pcrosemurgy — APPROVED** [review]
Thanks Dylan, looks sane to me :)

File: db-tests/src/test/java/TestTriggers.java
Comment (Twisol) [inline db-tests/src/test/java/TestTriggers.java]: Imports are a little unordered here -- this `org.apache` is sandwiched between two `org.junit`.

Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Addressed

File: db-tests/src/test/java/TestTriggers.java
Comment (patkenneally) [inline db-tests/src/test/java/TestTriggers.java:3]: 🔶 I think the class and the file should be named the same DatabaseTests.java and DatabaseTests. Or if there's a more apt naming that would work just as well.

File: db-tests/src/test/java/TestTriggers.java
Comment (patkenneally) [inline db-tests/src/test/java/TestTriggers.java]: 🔻 My IntelliJ is telling me that we should favor a static import of the classes we need from org.junit.jupiter.api.Assumptions rather than using the older org.junit.Assume. I think there are two instances in this PR.

File: db-tests/src/test/java/TestTriggers.java
Comment (patkenneally) [inline db-tests/src/test/java/TestTriggers.java]: 🔶 I'm interested to discuss this assertion choice. I may be missing something here but it seems like this is always true given the return type of the `res.getInt`. Could it be more descriptive to test something about the value like maybe check `newDatasetId > 0` or whatever we expect it to be?

**Review by patkenneally — DISMISSED** [review]
Legend:
🔻 **Issues** to address before merge
🔶 **Requests** that should not block merge, but should at least be discussed
🔵 **Recommendations** that can be skipped if desired, but should be acknowledged.
&nbsp;&nbsp;&nbsp;&nbsp; **Observations** that don't even need to be acknowledged

🔶 Are we able to move these tests into a java package path, something like merlin-database/src/test/java/gov/nasa/jpl/aerie/database ?

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java:3]: I second this ^

File: db-tests/build.gradle
Comment (kroffo) [inline db-tests/build.gradle]: This dependency is solely so we can use `Pair` right? I think it would probably be better to just make a record in place of the pairs we are using, primarily for increased visibility into what the pairs are, and secondly to avoid increasing the size of dependencies.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java]: Could use a record like `PlanDatasetRecord` to represent these instead of a `Pair`. Looking at other modules such as `merlin-server`, we do use a `model/` package for such purposes in our testing, so we should do that here too.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java]: Replacing the pair with a record would also make this field more clearly a record of data.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java]: A better check might be to call `assertFalse(res.wasNull());`

`res.getInt()` returns 0 if the value was null, which gets by the current check without issue: https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html#getInt(int)

However, we can check that the value `getInt()` determined it's return value from was not null using `res.wasNull()`: https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html#wasNull()

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java:672]: If we go with the `res.wasNull()` check on the ID above, we could do the same here. However, I also wonder if we should make it a separate test to check the revision. I think when we are testing more than one thing it indicates the presence of two tests in one. If one thing is wrong, we can't tell if it's truly one thing or both things that are broken.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java]: All of this is setting up initial conditions for the actual test which is that an inserted plan_dataset with a dataset_id should automatically compute the offset_from_plan_start.

It would be better to have the initial state before this state have two plans and one plan_dataset already in the database, and have the required information (start times, and offset) stored in variables so you don't have to read the database to get them. You're already doing this anyway with a bit of other information, it won't hurt to add some more initial state.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java]: Would you mind storing the result of the computation in a variable before this so that the assertion is more clear at a glance?

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [inline db-tests/src/test/java/TestTriggers.java]: Would you mind making all function parameters `final` unless there's a good reason not to? We try to default to `final` as much as possible unless there's a good reason not to. You can also set your editor up to flag this for you if you would like.

**Review by kroffo — CHANGES_REQUESTED** [review]
Good work on the tests! Some points to address, but overall great stuff!

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java:672]: My thinking here was to just test that the fields of the dataset are the defaults - hence the revision assertion here.

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Order of events here:
get currently existing plan start time
get the offset for the associated plan from the plan_dataset currently existing
insert a new plan with a different start time
grab the start time from the new plan
insert a new plan_dataset, which fires the trigger, and get the offset from it
assert the new offset is correct

Your suggestion is to add the second plan with the different time as part of test setup? I did it directly as part of this test since it's specific to this test and not really a shared setup. We can definitely have an additional beforeEach specific to this test, but that seemed like misdirection and overhead.

File: db-tests/build.gradle
Comment (dyst5422) [reply db-tests/build.gradle]: We can do that. I was honestly surprised that tuple types weren't part of the language, but a bolt on class.

Comment (Twisol) [reply db-tests/build.gradle]: Java was, and still is in some ways, a very restricted language. 😅 We've only recently got some more niceties.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java]: I wish we could have a `@Before` annotation that applies only to this specific test... I checked before I wrote my above comment and it's not included in the API we are using, so that's not an option.

Maybe you could just move the setup to a separate function just above this and call it with a comment explaining that the function just sets up initial state for the test? Then if a failure happens in the initialization, at least it's clear from the stack trace it's in the initialization function, and not the test itself.

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java:3]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/build.gradle
Comment (dyst5422) [reply db-tests/build.gradle]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: Done

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java]: Sorry to be so picky (though at the same times, we _should_ hold ourselves to high standards) but the name `PlanDatasetId` for the record seems misleading, as it's not the ID of a PlanDataset, but rather a record of a PlanDataset, which is at its most basic form just two IDs, the association of a plan ID with a Dataset ID. If you have opposing thoughts I'm happy to hear you out, but I do think we should rename it to `PlanDatasetRecord` (to be clear, not because it's a Java record, but because it's a literal record of data).

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java
Comment (kroffo) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java]: The datasetID should be included in the `PlanDatasetId` record, which is not just holding a plan ID (see comment about renaming the record). In fact, the field `datasetId` is never used throughout the tests, and can be removed (see comment on `beforeEach()` method)

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java
Comment (kroffo) [inline db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java]: The insertion of a plan dataset automatically creates the dataset, so you shouldn't explicitly create a dataset unless you _also_ want one that isn't associated with the plan, but you aren't using it in your tests so I don't think that is the case.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java]: Resolving, but there are new comments to address in this area as well.

File: db-tests/src/test/java/TestTriggers.java
Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java:672]: Yeah, it doesn't seem like a big deal to me, just thinking about the principles more than the specific situation. I'm personally fine with checking the two things here, but if it ever does come up as an issue we should probably split it.

File: db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java
Comment (dyst5422) [reply db-tests/src/test/java/gov/nasa/jpl/aerie/database/DatabaseTests.java]: Agreed. I'm going to leave the dataset insert method in there for possible future use cases, but just remove the call and assignment

File: db-tests/src/test/java/TestTriggers.java
Comment (dyst5422) [reply db-tests/src/test/java/TestTriggers.java]: I don't think the benefits outweigh the extra code necessary for that. Since we need the return value of a bunch of that "initialization", in addition to the new plan id, we'd have an additional method (not too bad here) and an additional record to get that data out of the scope of the function that really only exists to get that data out of the function - which is again a very specific initialization function for this particular test.

Comment (kroffo) [reply db-tests/src/test/java/TestTriggers.java]: Well, since there's no elegant solution here I'm fine with keeping it how it is, so I'll resolve.

**Review by kroffo — APPROVED** [review]

**Pull request merged**
